compiler/src/dotty/tools/dotc/typer/Nullables.scala

```diff
@@ -67,15 +65,18 @@ object Nullables:
       if this.isEmpty then that
       else if that.isEmpty then this
       else NotNullInfo(
-        this.asserted.union(that.asserted).diff(that.retracted),
-        this.retracted.union(that.retracted).diff(that.asserted))
+        this.asserted.diff(that.retracted).union(that.asserted),
+        this.retracted.union(that.retracted))
 
     /** The alternative path combination with another not-null info. Used to merge
      *  the nullability info of the two branches of an if.
      */
     def alt(that: NotNullInfo): NotNullInfo =
       NotNullInfo(this.asserted.intersect(that.asserted), this.retracted.union(that.retracted))
 
+    def withRetracted(that: NotNullInfo): NotNullInfo =
```

Comment:
Here it would be useful to have a comment explaining when this function is intended to be used. The name explains what the function does, but the idea is to say why one would want to do that. (Compare with the `alt` function, whose name says when you want it.) It looks like the use case is similar to `alt`, but in a case when one branch is known to not complete execution normally (e.g. an infinite loop), only exceptionally. We can say that in the comment and/or consider incorporating that into the name somehow.

This also raises the question of what to do when both alternatives cannot complete normally. Theoretically one would want to `assert` all variables, but we don't have a way to do that in the implementation. Perhaps that case is not common enough in practice to care about.

Reply:
I don't understand why we want to `assert` all variables if none of the alternatives complete normally?  Do you have an example?

Reply:
In general the analysis proves the following condition: in every execution trace leading to a given program point $p$, variable `x` is not null. When the code before $p$ does not terminate, there are no traces leading to $p$ so the condition is true for every variable. If there were an instruction derefencing `x` at $p$, that instruction would never execute, so it would be safe to assume that `x` is not null.

The specific case of `withRetracted` handles something like `alt`, but where the `that` branch does not complete normally. It is defined as `<a1,r1>.withRetracted(<a2,r2>) = <a1, r1 union r2>`.  `alt` is defined as `<a1,r1>.alt(<a2,r2>) = <a1 intersect a2, r1 union r2>`. If we could represent the universal set in `a2`, then `a1 intersect a2` would be `a1`, so we could implement the equivalent of `withRetracted` using `alt`.

I don't know whether actual use cases in practice justify the code complexity of adding a representation of the universal set to the implementation, but it is cleaner in theory.

If we don't do that, then the comment/name of `withRetracted` should suggest that it is like `alt`, but for the case where `that` cannot complete normally.